Feature/ecflow suite generator

KyleNevins-NOAA opened the pull request:
**Description**
This PR is to merge in the workflow generator application that uses a YAML file to create suite definition files for ecflow. 

**Type of change**
- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

**How Has This Been Tested?**
To test: 
1. Change into the workflow directory: ` cd global-workflow/workflow_generator`
1. Ensure the YAML file is configured with the suites/tasks that are wanted. 
1. Run the utility: `python3 setup_workflow.py --expdir ../parm/config`

** Remaining items **
1. Additional documentation
1. Code documentation
1. Test suite with existing production ecflow definition file.

KyleNevins-NOAA committed:
Initial commit for the setup_workflow_ecflow

Currently working with the ecflow base configurations. It loads in
the tasks, families, and various other nodes, allowing for
different suite hours to be created.

KyleNevins-NOAA committed:
Initial commit for the setup_workflow_ecflow

Currently working with the ecflow base configurations. It loads in
the tasks, families, and various other nodes, allowing for
different suite hours to be created.

KyleNevins-NOAA committed:
Added in the GFSApp definition file

After coordinating with Rahul, Kate, Walter, and Lin, the result
was a slimmed down version of the GFSApp for forecast only definition
file. Committed here for use with unittests and reference

KyleNevins-NOAA committed:
Added in the GFSApp definition file

After coordinating with Rahul, Kate, Walter, and Lin, the result
was a slimmed down version of the GFSApp for forecast only definition
file. Committed here for use with unittests and reference

KyleNevins-NOAA committed:
Reworked structure for the generator

After the meeting on 2/18, the generator was reworked to use
directory traversal to build the suites. This content was presented
on 3/1 but ultimately determined that the family structure should
be within the YAML so this is a working version but will be redone
to meet the new requirements.

KyleNevins-NOAA committed:
Reworked structure for the generator

After the meeting on 2/18, the generator was reworked to use
directory traversal to build the suites. This content was presented
on 3/1 but ultimately determined that the family structure should
be within the YAML so this is a working version but will be redone
to meet the new requirements.

KyleNevins-NOAA committed:
Updated scripts for new build structure

This is the new setup and definitions setup using the new ecflow
build structure vetted by the team. Still requires inclusion of
triggers and events.

KyleNevins-NOAA committed:
Updated scripts for new build structure

This is the new setup and definitions setup using the new ecflow
build structure vetted by the team. Still requires inclusion of
triggers and events.

KyleNevins-NOAA committed:
Restructured classes based one new requirements

Completion of the work by 3/11/22. Given the new requirements last
week, this refactors the ecflow_setup script to handle the configuration
file setup and then all the defintion pieces are handled in the
ecflow_definitions. This also manages the looping structures much
better and adds in some exceptions.

KyleNevins-NOAA committed:
Restructured classes based one new requirements

Completion of the work by 3/11/22. Given the new requirements last
week, this refactors the ecflow_setup script to handle the configuration
file setup and then all the defintion pieces are handled in the
ecflow_definitions. This also manages the looping structures much
better and adds in some exceptions.

KyleNevins-NOAA committed:
Removal of old files

There were files originally included with the first release that are
no longer needed. This removes them.

KyleNevins-NOAA committed:
Removal of old files

There were files originally included with the first release that are
no longer needed. This removes them.

KyleNevins-NOAA committed:
Added in Edits with lists

Added in the list capabilities for the inventory for edits.

KyleNevins-NOAA committed:
Added in Edits with lists

Added in the list capabilities for the inventory for edits.

KyleNevins-NOAA committed:
Implemented the event and task lists trigger loops still broken

Updated the logic for the tasks and the events to loop through
lists but the triggers logic is still broken if the events
are looped.

KyleNevins-NOAA committed:
Implemented the event and task lists trigger loops still broken

Updated the logic for the tasks and the events to loop through
lists but the triggers logic is still broken if the events
are looped.

KyleNevins-NOAA committed:
Updated Trigger logic to include event loops

The event loops previously errored out due to the int being cast
as strings. This commit fixes that up.

KyleNevins-NOAA committed:
Updated Trigger logic to include event loops

The event loops previously errored out due to the int being cast
as strings. This commit fixes that up.

KyleNevins-NOAA committed:
Initial repeat logic implementation

This generates a date repeat logic and basic time but doesn't work
properly due to time being limited to a 24 hour time stamp.

KyleNevins-NOAA committed:
Initial repeat logic implementation

This generates a date repeat logic and basic time but doesn't work
properly due to time being limited to a 24 hour time stamp.

KyleNevins-NOAA committed:
Added in multiple repeat options

Now uses cron for abnormal times over a day, a time if the start
and end is within the same day and date if it is looping through
days

KyleNevins-NOAA committed:
Added in multiple repeat options

Now uses cron for abnormal times over a day, a time if the start
and end is within the same day and date if it is looping through
days

KyleNevins-NOAA committed:
Dates with repeats added.

This commit was done after the demonstration on 3/29 to the team.
The next steps are to change out the env. for ${} syntax and then
also start trying to put together a suite definition from current
production to identify any possible errors.

Additional tasks are:
- Documentation
- Identify a way to use previously defined variables within the
manifest with $()

KyleNevins-NOAA committed:
Dates with repeats added.

This commit was done after the demonstration on 3/29 to the team.
The next steps are to change out the env. for ${} syntax and then
also start trying to put together a suite definition from current
production to identify any possible errors.

Additional tasks are:
- Documentation
- Identify a way to use previously defined variables within the
manifest with $()

KyleNevins-NOAA committed:
First README.md file

Added in a readme file for use of the workflow object.

KyleNevins-NOAA committed:
First README.md file

Added in a readme file for use of the workflow object.

KyleNevins-NOAA committed:
Merge branch 'feature/ecflow_suite_generator' of noaa-github:KyleNevins-NOAA/global-workflow into feature/ecflow_suite_generator

KyleNevins-NOAA committed:
Addition of the production definition and add ons

This change adds in the prod.yml file which is the start of the production
definition file and the creation of multiple suites. In addition there have
been fixes like the addition of the defstatus.

KyleNevins-NOAA committed:
Updated production YAML with OR logic

Added in the ability for the triggers to have an "or" addition
and also updated the production YAML further.

WalterKolczynski-NOAA reviewed (changes requested):
Looks good in general, except it is lacking in documentation and doesn't appear to follow PEP8. I don't think it is too far off PEP8 and I haven't run a linter yet, but I know there's supposed to be a space after commas, for example.

KyleNevins-NOAA committed:
Updated Documentation

Added additional sections to the README as well as updated in line
code for the main script.

KyleNevins-NOAA committed:
Updated PEP8 and Documentation

Added documentation to the setup_workflow and ecflow_setup modules
and also made some adjustments based on flake8's recommendations.

KyleNevins-NOAA commented:
> Looks good in general, except it is lacking in documentation and doesn't appear to follow PEP8. I don't think it is too far off PEP8 and I haven't run a linter yet, but I know there's supposed to be a space after commas, for example.

Have made some adjustments to two of the modules but more documentation/pep8/readme updates incoming.

KyleNevins-NOAA committed:
Updated code documentation and PEP8 compliance

Added in more code documentation and PEP 8 compliance.

KyleNevins-NOAA committed:
Updated README and documentation

Added the documentation for the various nodes along with notes
about the ranges and lists to the readme. Also added in docs
to the code following PEP8 style.

KyleNevins-NOAA commented:
Added in additional documentation within the code, I believe all methods, classes, and program descriptions are now documented. Also updated the readme.

WalterKolczynski-NOAA commented on workflow_generator/ecflow_setup/ecflow_definitions.py:
```diff
@@ -117,95 +376,179 @@ def add_repeat(self,repeat,parent=None):
         if operand is not None:
-            addTrigger = ecflow.Trigger(addTrigger.get_expression(),operand)
-        self.ecfnodes[parent].add(addTrigger)
+            addTrigger = ecflow.Trigger(addtrigger.get_expression(), operand)
```
left side is still using camelcase (should all these be snake case, i.e. add_trigger)?

KyleNevins-NOAA replied:
Using snake case now. If you see any others, please let me know and I'll adjust.

WalterKolczynski-NOAA commented on workflow_generator/ecflow_setup/ecflow_definitions.py:
```diff
@@ -423,6 +994,17 @@ def set_max_value(self,range_token):
                 sys.exit(1)
 
     def set_initial_max_value(self,range_token):
```
Missing space after comma

KyleNevins-NOAA replied:
Shoot, thought I got them all, was bound to miss one or two. Thanks, fixed it up.

WalterKolczynski-NOAA commented on workflow_generator/ecflow_setup/ecflow_definitions.py:
```diff
@@ -423,6 +994,17 @@ def set_max_value(self,range_token):
                 sys.exit(1)
 
     def set_initial_max_value(self,range_token):
```
Missing space after comma

WalterKolczynski-NOAA commented on workflow_generator/ecflow_setup/ecflow_definitions.py:
```diff
@@ -389,27 +932,55 @@ def __init__(self,ecfItem):
 
     def get_name(self):
+        """
+        Returns the name of the node.
```
I know it seems redundant/silly, but this and the next few methods should have a Returns section for consistency (especially once we start using a documentation generator to turn these blocks into documentation)

KyleNevins-NOAA replied:
Added in a RETURNS or PARAMETERS section to all of the methods. Added value NONE if there wasn't actually something for that section.

WalterKolczynski-NOAA commented on workflow_generator/ecflow_setup/ecflow_definitions.py:
```diff
@@ -389,27 +932,55 @@ def __init__(self,ecfItem):
 
     def get_name(self):
+        """
+        Returns the name of the node.
```
I know it seems redundant/silly, but this and the next few methods should have a Returns section for consistency (especially once we start using a documentation generator to turn these blocks into documentation)

WalterKolczynski-NOAA commented on workflow_generator/ecflow_setup/ecflow_definitions.py:
```diff
@@ -771,3 +1759,4 @@ class RangeError(Error):
 class ConfigurationError(Error):
     """Raised when there is an error in the configuration file."""
     pass
+7
```
Erroneous 7?

KyleNevins-NOAA replied:
Yeah, not sure how that got there. Lucky number 7 has been removed.

WalterKolczynski-NOAA commented on workflow_generator/ecflow_setup/ecflow_definitions.py:
```diff
@@ -771,3 +1759,4 @@ class RangeError(Error):
 class ConfigurationError(Error):
     """Raised when there is an error in the configuration file."""
     pass
+7
```
Erroneous 7?

WalterKolczynski-NOAA reviewed (changes requested):
Descriptions look great. For consistency, I think we should have a Parameters and Returns for every method, even if we just put None in the category.

KyleNevins-NOAA committed:
PEP8 and documentation updates with one bug fix

Based on the PR review, this commit adds in a lot of the RETURN
and PARAMETER sections even if the value of them is none. Some of
the sections were missed during the documentation phase. Also,
some of the camelcase was fixed to use snake case.

Also, there was a bug fix for the events range function. It was
using a bad variable and so it didn't actually add the event.
